org.apache.sshd.client.global.OpenSshHostKeysHandler should ignore unsupported host keys

### Version

2.9.2

### Bug description

When receiving `hostkeys-00@openssh.com` request, client by default uses `org.apache.sshd.client.global.OpenSshHostKeysHandler` to process it. This handler's `handleHostKeys` method does nothing but logging keys when debug is enabled.

But there is no exception handling in parent `org.apache.sshd.common.global.org.apache.sshd.common.global.AbstractOpenSshHostKeysHandler` class.

Therefore if the host sends two keys (one RSA and one EdDSA for example), but the client only supports RSA, the exception is being propagated from the `org.apache.sshd.common.global.AbstractOpenSshHostKeysHandler#process` method and 
the host's request is processed as failed and the WARN message is generated in logs.

### Actual behavior

Example:
`WARN 40340 --- []-nio2-thread-3] o.a.s.c.session.ClientConnectionService  : globalRequest(ClientConnectionService[ClientSessionImpl[test@localhost/127.0.0.1:2222]])[hostkeys-00@openssh.com, want-reply=false] failed (SshException) to process: EdDSA provider not supported`

### Expected behavior

No warns are logged, and only the supported keys are handled by `org.apache.sshd.client.global.OpenSshHostKeysHandler#handleHostKeys`

### Relevant log output

```Shell
WARN 40340 --- []-nio2-thread-3] o.a.s.c.session.ClientConnectionService  : globalRequest(ClientConnectionService[ClientSessionImpl[test@localhost/127.0.0.1:2222]])[hostkeys-00@openssh.com, want-reply=false] failed (SshException) to process: EdDSA provider not supported

org.apache.sshd.common.SshException: EdDSA provider not supported
	at org.apache.sshd.common.util.buffer.Buffer.getRawPublicKey(Buffer.java:544) ~[sshd-common-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.util.buffer.Buffer.getPublicKey(Buffer.java:519) ~[sshd-common-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.global.AbstractOpenSshHostKeysHandler.process(AbstractOpenSshHostKeysHandler.java:73) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.global.AbstractOpenSshHostKeysHandler.process(AbstractOpenSshHostKeysHandler.java:38) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.session.helpers.AbstractConnectionService.globalRequest(AbstractConnectionService.java:868) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.session.helpers.AbstractConnectionService.process(AbstractConnectionService.java:486) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.session.helpers.CurrentService.process(CurrentService.java:109) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.session.helpers.AbstractSession.doHandleMessage(AbstractSession.java:591) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.session.helpers.AbstractSession.lambda$handleMessage$0(AbstractSession.java:522) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.util.threads.ThreadUtils.runAsInternal(ThreadUtils.java:68) ~[sshd-common-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.session.helpers.AbstractSession.handleMessage(AbstractSession.java:521) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.session.helpers.AbstractSession.decode(AbstractSession.java:1639) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.session.helpers.AbstractSession.messageReceived(AbstractSession.java:482) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.session.helpers.AbstractSessionIoHandler.messageReceived(AbstractSessionIoHandler.java:64) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.io.nio2.Nio2Session.handleReadCycleCompletion(Nio2Session.java:407) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.io.nio2.Nio2Session$1.onCompleted(Nio2Session.java:380) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.io.nio2.Nio2Session$1.onCompleted(Nio2Session.java:375) ~[sshd-core-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.io.nio2.Nio2CompletionHandler.lambda$completed$0(Nio2CompletionHandler.java:38) ~[sshd-core-2.9.2.jar:2.9.2]
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318) ~[na:na]
	at org.apache.sshd.common.io.nio2.Nio2CompletionHandler.completed(Nio2CompletionHandler.java:37) ~[sshd-core-2.9.2.jar:2.9.2]
	at java.base/sun.nio.ch.Invoker.invokeUnchecked(Invoker.java:129) ~[na:na]
	at java.base/sun.nio.ch.Invoker$2.run(Invoker.java:221) ~[na:na]
	at java.base/sun.nio.ch.AsynchronousChannelGroupImpl$1.run(AsynchronousChannelGroupImpl.java:113) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:840) ~[na:na]
Caused by: java.security.NoSuchAlgorithmException: EdDSA provider not supported
	at org.apache.sshd.common.util.security.SecurityUtils.generateEDDSAPublicKey(SecurityUtils.java:627) ~[sshd-common-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.util.buffer.keys.ED25519BufferPublicKeyParser.getRawPublicKey(ED25519BufferPublicKeyParser.java:46) ~[sshd-common-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.util.buffer.keys.BufferPublicKeyParser$2.getRawPublicKey(BufferPublicKeyParser.java:102) ~[sshd-common-2.9.2.jar:2.9.2]
	at org.apache.sshd.common.util.buffer.Buffer.getRawPublicKey(Buffer.java:542) ~[sshd-common-2.9.2.jar:2.9.2]
	... 25 common frames omitted
```


### Other information

_No response_

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -29,6 +29,8 @@
 ## Bug Fixes
 
 * [GH-428/GH-392](https://github.com/apache/mina-sshd/issues/428) SCP client fails silently when error signalled due to missing file or lacking permissions
+* [GH-434](https://github.com/apache/mina-sshd/issues/434) Ignore unknown key types from agent or in OpenSSH host keys extension
+
 
 ## New Features
 

--- a/sshd-common/src/main/java/org/apache/sshd/common/util/buffer/Buffer.java
+++ b/sshd-common/src/main/java/org/apache/sshd/common/util/buffer/Buffer.java
@@ -537,11 +537,14 @@ public abstract class Buffer implements Readable {
     public PublicKey getPublicKey(BufferPublicKeyParser<? extends PublicKey> parser) throws SshException {
         int ow = wpos();
         int len = ensureAvailable(getInt());
-        wpos(rpos() + len);
+        int afterKey = rpos() + len;
+        wpos(afterKey);
         try {
             return getRawPublicKey(parser);
         } finally {
             wpos(ow);
+            // Skip this key, even if the parser failed.
+            rpos(afterKey);
         }
     }
 

--- a/sshd-core/src/main/java/org/apache/sshd/agent/common/AbstractAgentProxy.java
+++ b/sshd-core/src/main/java/org/apache/sshd/agent/common/AbstractAgentProxy.java
@@ -92,13 +92,25 @@ public abstract class AbstractAgentProxy extends AbstractLoggingBean implements 
         List<SimpleImmutableEntry<PublicKey, String>> keys = new ArrayList<>(nbIdentities);
         boolean debugEnabled = log.isDebugEnabled();
         for (int i = 0; i < nbIdentities; i++) {
-            PublicKey key = buffer.getPublicKey();
-            String comment = buffer.getString();
-            if (debugEnabled) {
-                log.debug("getIdentities() key type={}, comment={}, fingerprint={}",
-                        KeyUtils.getKeyType(key), comment, KeyUtils.getFingerPrint(key));
+            PublicKey key = null;
+            String error = null;
+            try {
+                key = buffer.getPublicKey();
+            } catch (SshException e) {
+                error = e.getLocalizedMessage();
             }
-            keys.add(new SimpleImmutableEntry<>(key, comment));
+            String comment = buffer.getString();
+            if (key == null) {
+                if (log.isTraceEnabled()) {
+                    log.trace("getIdentities() ignoring an unknown key type; comment={}; error={}", comment, error);
+                }
+            } else {
+                if (debugEnabled) {
+                    log.debug("getIdentities() key type={}, comment={}, fingerprint={}", KeyUtils.getKeyType(key), comment,
+                            KeyUtils.getFingerPrint(key));
+                }
+                keys.add(new SimpleImmutableEntry<>(key, comment));
+            }
         }
 
         return keys;

--- a/sshd-core/src/main/java/org/apache/sshd/client/global/OpenSshHostKeysHandler.java
+++ b/sshd-core/src/main/java/org/apache/sshd/client/global/OpenSshHostKeysHandler.java
@@ -43,10 +43,12 @@ public class OpenSshHostKeysHandler extends AbstractOpenSshHostKeysHandler {
 
     public OpenSshHostKeysHandler() {
         super(REQUEST);
+        setIgnoreInvalidKeys(true);
     }
 
     public OpenSshHostKeysHandler(BufferPublicKeyParser<? extends PublicKey> parser) {
         super(REQUEST, parser);
+        setIgnoreInvalidKeys(true);
     }
 
     @Override

--- a/sshd-core/src/main/java/org/apache/sshd/common/global/AbstractOpenSshHostKeysHandler.java
+++ b/sshd-core/src/main/java/org/apache/sshd/common/global/AbstractOpenSshHostKeysHandler.java
@@ -24,6 +24,7 @@ import java.util.Collection;
 import java.util.LinkedList;
 import java.util.Objects;
 
+import org.apache.sshd.common.SshException;
 import org.apache.sshd.common.config.keys.KeyUtils;
 import org.apache.sshd.common.session.ConnectionService;
 import org.apache.sshd.common.session.Session;
@@ -38,6 +39,8 @@ import org.apache.sshd.common.util.buffer.keys.BufferPublicKeyParser;
 public abstract class AbstractOpenSshHostKeysHandler extends AbstractConnectionServiceRequestHandler {
     private final String request;
     private final BufferPublicKeyParser<? extends PublicKey> parser;
+
+    private boolean ignoreInvalidKeys;
 
     protected AbstractOpenSshHostKeysHandler(String request) {
         this(request, BufferPublicKeyParser.DEFAULT);
@@ -70,18 +73,36 @@ public abstract class AbstractOpenSshHostKeysHandler extends AbstractConnectionS
         if (p != null) {
             boolean debugEnabled = log.isDebugEnabled();
             while (buffer.available() > 0) {
-                PublicKey key = buffer.getPublicKey(p);
-                if (debugEnabled) {
-                    log.debug("process({})[{}] key type={}, fingerprint={}",
-                            connectionService, request, KeyUtils.getKeyType(key), KeyUtils.getFingerPrint(key));
+                PublicKey key = null;
+                try {
+                    key = buffer.getPublicKey(p);
+                } catch (SshException e) {
+                    if (!isIgnoreInvalidKeys()) {
+                        throw e;
+                    }
+                    if (log.isTraceEnabled()) {
+                        log.trace("process({})[{}] received an unsupported key", connectionService, request, e);
+                    }
                 }
                 if (key != null) {
+                    if (debugEnabled) {
+                        log.debug("process({})[{}] key type={}, fingerprint={}", connectionService, request,
+                                KeyUtils.getKeyType(key), KeyUtils.getFingerPrint(key));
+                    }
                     keys.add(key);
                 }
             }
         }
 
         return handleHostKeys(connectionService.getSession(), keys, wantReply, buffer);
+    }
+
+    protected void setIgnoreInvalidKeys(boolean ignore) {
+        ignoreInvalidKeys = ignore;
+    }
+
+    protected boolean isIgnoreInvalidKeys() {
+        return ignoreInvalidKeys;
     }
 
     protected abstract Result handleHostKeys(

--- a/sshd-core/src/test/java/org/apache/sshd/common/global/OpenSshHostKeysHandlerTest.java
+++ b/sshd-core/src/test/java/org/apache/sshd/common/global/OpenSshHostKeysHandlerTest.java
@@ -1,0 +1,110 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.sshd.common.global;
+
+import java.security.GeneralSecurityException;
+import java.security.KeyPairGenerator;
+import java.security.PublicKey;
+import java.util.Collection;
+
+import org.apache.sshd.common.SshException;
+import org.apache.sshd.common.config.keys.KeyUtils;
+import org.apache.sshd.common.session.ConnectionService;
+import org.apache.sshd.common.session.Session;
+import org.apache.sshd.common.util.buffer.Buffer;
+import org.apache.sshd.common.util.buffer.ByteArrayBuffer;
+import org.apache.sshd.util.test.BaseTestSupport;
+import org.apache.sshd.util.test.NoIoTestCase;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+/**
+ * @author <a href="mailto:dev@mina.apache.org">Apache MINA SSHD Project</a>
+ */
+@Category(NoIoTestCase.class)
+@RunWith(MockitoJUnitRunner.class)
+public class OpenSshHostKeysHandlerTest extends BaseTestSupport {
+
+    @Mock
+    private ConnectionService connectionService;
+
+    private PublicKey key;
+    private Buffer buffer;
+
+    public OpenSshHostKeysHandlerTest() {
+        super();
+    }
+
+    @Before
+    public void prepareBuffer() throws Exception {
+        // Create an RSA key
+        key = KeyPairGenerator.getInstance("RSA").generateKeyPair().getPublic();
+        // Serialize it twice to a buffer, but insert a fake item in between
+        buffer = new ByteArrayBuffer();
+        buffer.putPublicKey(key);
+        buffer.putUInt(34);
+        buffer.putString("unknown"); // Fake key type; 7 + 4 bytes length
+        buffer.putString("followed by garbage"); // 19 + 4
+        buffer.putPublicKey(key);
+    }
+
+    @Test
+    public void clientIgnoresUnknownKeys() throws Exception {
+        boolean[] handlerCalled = { false };
+        org.apache.sshd.client.global.OpenSshHostKeysHandler handler
+                = new org.apache.sshd.client.global.OpenSshHostKeysHandler() {
+                    @Override
+                    protected Result handleHostKeys(
+                            Session session, Collection<? extends PublicKey> keys, boolean wantReply,
+                            Buffer buffer) throws Exception {
+                        handlerCalled[0] = true;
+                        assertEquals("Unexpected number of keys", 2, keys.size());
+                        for (PublicKey k : keys) {
+                            assertTrue("Unexpected public key", KeyUtils.compareKeys(key, k));
+                        }
+                        return Result.Replied;
+                    }
+                };
+        handler.process(connectionService, org.apache.sshd.client.global.OpenSshHostKeysHandler.REQUEST, false, buffer);
+        assertTrue("Handler should have been called", handlerCalled[0]);
+    }
+
+    @Test
+    public void serverThrowsOnUnknownKeys() throws Exception {
+        boolean[] handlerCalled = { false };
+        org.apache.sshd.server.global.OpenSshHostKeysHandler handler
+                = new org.apache.sshd.server.global.OpenSshHostKeysHandler() {
+                    @Override
+                    protected Result handleHostKeys(
+                            Session session, Collection<? extends PublicKey> keys, boolean wantReply,
+                            Buffer buffer) throws Exception {
+                        handlerCalled[0] = true;
+                        return Result.Replied;
+                    }
+                };
+        SshException e = assertThrows(SshException.class, () -> handler.process(connectionService,
+                org.apache.sshd.server.global.OpenSshHostKeysHandler.REQUEST, false, buffer));
+        assertFalse("Handler should not have been called", handlerCalled[0]);
+        assertTrue("Expected exception cause", e.getCause() instanceof GeneralSecurityException);
+    }
+}

